akka-stream/src/main/scala/akka/stream/impl/io/SslTlsCipherActor.scala

```diff
@@ -299,6 +299,8 @@ private[akka] class SslTlsCipherActor(settings: ActorMaterializerSettings, sslCo
       } catch {
         case ex: SSLException ⇒
           if (tracing) log.debug(s"SSLException during doUnwrap: $ex")
+          fail(ex, closeTransport = false)
+          engine.closeInbound()
```

Comment:
It's not redundant, in that the provided test fails without the `engine.closeInbound()` call. The connection is never closed, and the reaper detects the dangling actors and fails the test.

*Why* this is true is not clear to me. Reading the [documentation](https://docs.oracle.com/javase/7/docs/api/javax/net/ssl/SSLEngine.html) for `SSLEngine`:

> A close can be initiated by one of: an SSLException, an inbound closure handshake message, or one of the close methods. In all cases, closure handshake messages are generated by the engine, and wrap() should be repeatedly called until the resulting SSLEngineResult's status returns "CLOSED", or isOutboundDone() returns true. All data obtained from the wrap() method should be sent to the peer.

So it sounds like if `doUnwrap()` raises `SSLException`, it should have already called `closeInbound()`. But that's clearly *not* the case, because the `closeInbound()` call is not a no-op. I'm not sure if this is an `SSLEngine` bug, bad / outdated `SSLEngine` documentation, or (most likely) me not fully understanding `SSLEngine`.

It's possible that some `SSLException`s are not considered "fatal", and thus will not close the inbound stream. I can't find any reference to this possibility in `SSLEngine` documentation.

I am frankly not convinced that my own closure logic is completely correct. `SSLEngine` is a complicated beast. But it does pass the test now, which means it is at least correct in cases where it was not before.